### DISCUSSION TITLE
Order by doesn't work on embedded field (Mongodb)

### DISCUSSION THREAD

**Discussion by Trobyss** [mikro-orm/mikro-orm#6002] [Q&A]
### Describe the bug

Hey ! 

I'm trying to sort on embedded field but got these error:

```
DriverException: Invalid sort direction: {"date":{"arrival":"ASC"}}
 at MongoExceptionConverter.convertException (/app/node_modules/@mikro-orm/core/platforms/ExceptionConverter.js:8:16)
 at MongoExceptionConverter.convertException (/app/node_modules/@mikro-orm/mongodb/MongoExceptionConverter.js:17:22)
 at MongoDriver.convertException (/app/node_modules/@mikro-orm/core/drivers/DatabaseDriver.js:345:54)
 at /app/node_modules/@mikro-orm/core/drivers/DatabaseDriver.js:349:24
 at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
 at async MongoDriver.find (/app/node_modules/@mikro-orm/mongodb/MongoDriver.js:49:21)
 at async MongoEntityManager.find (/app/node_modules/@mikro-orm/core/EntityManager.js:144:25)
 at async Promise.all (index 0)
 at async MongoEntityManager.findAndCount (/app/node_modules/@mikro-orm/core/EntityManager.js:383:35)
 at async MongoEntityManager.findByCursor (/app/node_modules/@mikro-orm/core/EntityManager.js:449:35)
```

I notice strange behavior:
It work's with flatten key.
```ts
        orderBy: {
          'details.date.arrival': query.sort,
        }
```

But didn't works with `first` option (I'm trying to sort with cursor)
```ts
      {
        first: 10,
        orderBy: {
          'details.date.arrival': query.sort,
        }
      }
```
The query generated (we can see that order by isn't here):
```
 [query] db.getCollection('rides').find({ '$and': [ { 'details.date.arrival': { '$lte': ISODate('2024-09-20T15:13:30.758Z') }, 'details.date.departure': { '$gte': ISODate('2024-09-05T15:13:30.758Z') } } ] }, {}).limit(11).toArray();
```

As far as I debugged, I think that there is an issue here: https://github.com/mikro-orm/mikro-orm/blob/8b30ae2a301c2e147d90983f4eaa08eb722e118f/packages/core/src/utils/Cursor.ts#L185 => the properties for the fields is like `details_datearrival` but the key is `details.date.arrival`.

### Reproduction
```ts


import {
  Embeddable,
  Embedded,
  Entity,
  MikroORM,
  PrimaryKey,
  Property,
  QueryOrder,
} from '@mikro-orm/mongodb';
import { MongoMemoryServer } from 'mongodb-memory-server';

@Embeddable()
export class RideDetailsDateModel {
  @Property()
  arrival: Date;

  @Property()
  departure: Date;

  constructor(arrival: Date, departure: Date) {
    this.arrival = arrival;
    this.departure = departure;
  }
}

@Embeddable()
export class RideDetailsModel {
  @Embedded({ object: true })
  date: RideDetailsDateModel;
}

@Entity({ collection: 'rides' })
export class RideModel {
  @PrimaryKey()
  _id!: number;

  @Embedded({ object: true })
  details: RideDetailsModel;
}

let orm: MikroORM;
let mongodb: MongoMemoryServer;

beforeAll(async () => {
  mongodb = await MongoMemoryServer.create();
  orm = await MikroORM.init({
    allowGlobalContext: true, // only for testing
    clientUrl: `${mongodb.getUri()}mikro`,
    debug: ['query', 'query-params'],
    entities: [RideModel, RideDetailsModel, RideDetailsDateModel],
  });
  await orm.schema.refreshDatabase();
});

afterAll(async () => {
  await mongodb.stop();
  await orm.close(true);
});

test('basic CRUD example', async () => {
  orm.em.create(RideModel, {
    details: {
      date: { arrival: new Date(2024, 1, 2), departure: new Date(2024, 1, 1) },
    },
  });
  orm.em.create(RideModel, {
    details: {
      date: { arrival: new Date(2024, 1, 4), departure: new Date(2024, 1, 3) },
    },
  });
  orm.em.create(RideModel, {
    details: {
      date: { arrival: new Date(2024, 1, 6), departure: new Date(2024, 1, 5) },
    },
  });
  orm.em.create(RideModel, {
    details: {
      date: { arrival: new Date(2024, 1, 8), departure: new Date(2024, 1, 7) },
    },
  });

  await orm.em.flush();
  orm.em.clear();

  const rides = await orm.em.findByCursor(
    RideModel,
    {},
    {
      orderBy: {
        details: {
          date: {
            departure: QueryOrder.ASC,
          },
        },
      },
    },
  );

  expect(rides.items[0].details.date).toStrictEqual(new Date(2024, 1, 1));
});
```

### What driver are you using?

@mikro-orm/mongodb

### MikroORM version

6.3.8

### Node.js version

20

### Operating system

node:20 docker image

### Validations

- [x] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Discord](https://discord.gg/w8bjxFHS7X).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
Ups, got confused with your partially working workaround with dot paths, since that is not what you should be doing and will fail with inlined mode, and when trying it out I used an inlined embedded property, which works just fine. You are correct that with object embedded property it's failing the way you mentioned, I believe that is the main thing that needs to be solved. And yeah, I also completely missed your reproduction 🤦 I guess its too hot in here :P

**Reply by Trobyss** [reply]
Yeah I know that my workaround is really weird, It's more a test that a production workaround ahah

So the short quick fix I can made is pass my Embedded properties to inline ?

I think the issue is here: https://github.com/mikro-orm/mikro-orm/blob/8b30ae2a301c2e147d90983f4eaa08eb722e118f/packages/core/src/utils/Cursor.ts#L185

Thanks you for your quick reply :D

**Reply by B4nan** [reply]
This will help with the dot paths 2a9ffa3a3d58ae7cf01a644277aa15cee0967fb7

**Reply by Trobyss** [reply]
Wow so speed! 

I'm going to try this later today and I come back to you :)

Thanks!

**Reply by B4nan** [reply]
fyi the cursor pagination is still broken, this only fixed the ordering, not the cursor definition

**Reply by Trobyss** [reply]
Oh yep, did you plan to investigate / fix it soon or do you need some helps ?

**Reply by B4nan** [reply]
This is actually quite hard to support, I think I am 80% there, but it takes a lot of time.
